# Proportional Approval Voting implementation
Todo: When it gets slow, use counts of unique cvrs, not all cvrs

When that gets slow, come up with more efficient algorithm, e.g. A*, which is easily distributed

In [1]:
import pandas as pd
import collections
import itertools
import time

import pav

In [2]:
winner, cvrs, scores, df = pav.tally_pav("Westminster_Adams.csv", 3)

Plurality results in order
Anita_Seitz         4921
Shannon_Bird        4823
Maria_De_Cambra     3921
Mark_Clark          3500
David_DeMott        3126
Debbie_Bergamo      2668
Mike_Melvin         2112
Jason_Blanckaert    2026
Steve_Caulk         2026
Nathan_Pearce       1229
dtype: int64

12684 CVRs, 169 unique selections of candidates
Score is 10619.167 for plurality winners ['Anita_Seitz', 'Shannon_Bird', 'Maria_De_Cambra']

Proportional Approval Voting results

Top 10 scores:
('10799.500', ['Anita_Seitz', 'Mark_Clark', 'Shannon_Bird'])
('10619.167', ['Anita_Seitz', 'Maria_De_Cambra', 'Shannon_Bird'])
('10607.500', ['Anita_Seitz', 'Debbie_Bergamo', 'Shannon_Bird'])
('10588.833', ['Anita_Seitz', 'David_DeMott', 'Shannon_Bird'])
('10366.333', ['Maria_De_Cambra', 'Mark_Clark', 'Shannon_Bird'])
('10349.500', ['Anita_Seitz', 'Mike_Melvin', 'Shannon_Bird'])
('10260.667', ['Anita_Seitz', 'Jason_Blanckaert', 'Shannon_Bird'])
('10214.000', ['Anita_Seitz', 'Maria_De_Cambra', 'Mark_Clark'])
('

In [3]:
# Use this to test out new code after modifying the pav module
import importlib
importlib.reload(pav)

<module 'pav' from '/srv/s/approvalvoting/pr_voting_methods/pav.py'>

In [4]:
df.head()

Mark_Clark  Mike_Melvin  Maria_De_Cambra  Shannon_Bird  Steve_Caulk  \
0           0            0                0             0            0   
1           0            0                1             1            1   
2           0            0                0             0            0   
3           0            0                0             0            1   
4           0            0                0             1            0   

   Debbie_Bergamo  Anita_Seitz  Jason_Blanckaert  David_DeMott  Nathan_Pearce  
0               0            0                 0             0              0  
1               0            0                 0             0              0  
2               0            0                 0             0              0  
3               0            1                 1             0              0  
4               1            0                 0             1              0

In [5]:
winner, cvrs, scores, df = pav.tally_pav("Byers_SD_32J_Adams.csv", 4)

Plurality results in order
Tom_Thompson_III        81
Jerry_L_Sauer           75
Christopher_P_Cary      57
Julie_Smith             56
Donna_J_Sauer           54
Yvonne_M_Gerhardt       47
Jennifer_Simanovicki    36
Margaret_A_Holeman      23
dtype: int64

121 CVRs, 50 unique selections of candidates
Score is 182.333 for plurality winners ['Tom_Thompson_III', 'Jerry_L_Sauer', 'Christopher_P_Cary', 'Julie_Smith']

Proportional Approval Voting results

Top 10 scores:
('182.333', ['Donna_J_Sauer', 'Jerry_L_Sauer', 'Julie_Smith', 'Tom_Thompson_III'])
('182.333', ['Christopher_P_Cary', 'Jerry_L_Sauer', 'Julie_Smith', 'Tom_Thompson_III'])
('180.833', ['Jerry_L_Sauer', 'Julie_Smith', 'Tom_Thompson_III', 'Yvonne_M_Gerhardt'])
('179.583', ['Christopher_P_Cary', 'Jerry_L_Sauer', 'Tom_Thompson_III', 'Yvonne_M_Gerhardt'])
('179.333', ['Christopher_P_Cary', 'Donna_J_Sauer', 'Jerry_L_Sauer', 'Tom_Thompson_III'])
('177.583', ['Donna_J_Sauer', 'Jerry_L_Sauer', 'Tom_Thompson_III', 'Yvonne_M_Gerhardt'])

# Timing results
Takes about 12 seconds(?) for 20 candidates, 10 winners, 121 ballots

Uses nearly 3 GB of memory with 24 and 12

24 choose 12 = 2704156

In [71]:
timings = []

In [77]:
candidates = 12
winners = 6

In [78]:
possible_results = itertools.combinations(all_candidates + tuple(list(range(candidates - len(all_candidates)))), winners)

In [79]:
t1 = time.time()
scores = {}
for result in possible_results:
    resultset = frozenset(result)
    scores[resultset] = sum(utility(resultset, cvr) for cvr in cvrs)
timings.append("{} seconds for {} candidates, {} winners, {} ballots".format(time.time() - t1, candidates, winners, len(cvrs)))

In [80]:
timings

['1031.4427270889282 seconds for 20 candidates, 10 winners, 121 ballots',
 '5.157426595687866 seconds for 12 candidates, 6 winners, 12684 ballots']

In [285]:
# older results
timings

['0.005549907684326172 seconds for 8 candidates, 4 winners, 121 ballots',
 '0.0004897117614746094 seconds for 8 candidates, 4 winners, 121 ballots',
 '0.022046566009521484 seconds for 10 candidates, 5 winners, 121 ballots',
 '0.22809982299804688 seconds for 14 candidates, 7 winners, 121 ballots',
 '2.5786209106445312 seconds for 18 candidates, 9 winners, 121 ballots',
 '9.772108793258667 seconds for 20 candidates, 10 winners, 121 ballots',
 '39.48445129394531 seconds for 22 candidates, 11 winners, 121 ballots',
 '161.45553970336914 seconds for 24 candidates, 12 winners, 121 ballots',
 '4.725412607192993 seconds for 24 candidates, 5 winners, 121 ballots',
 '6.555292844772339 seconds for 30 candidates, 5 winners, 121 ballots']